iceoryx_hoofs/test/moduletests/test_static_lifetime_guard.cpp

```diff
@@ -0,0 +1,221 @@
+
+        // still the same instance as g_instance
+        EXPECT_EQ(instance.id, 1);
+        EXPECT_EQ(&instance, &g_instance);
+    }
+
+    // the implicit guard of g_instance prevents destruction
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+    EXPECT_EQ(g_instance.id, 1);
+}
+
+TEST_F(StaticLifetimeGuard_test, copyIncreasesLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6ab6396d-7c63-4626-92ed-c7f3ea67bbf1");
+    Foo::reset();
+    Guard guard;
+    {
+        EXPECT_EQ(Guard::count(), 2);
+        // NOLINTJUSTIFICATION ctor and dtor side effects are tested
+        // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+        Guard copy(guard);
+        EXPECT_EQ(Guard::count(), 3);
+    }
+    EXPECT_EQ(Guard::count(), 2);
+
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, moveIncreasesLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "32a2fdbf-cb02-408c-99a3-373aa66b2764");
+    Foo::reset();
+    Guard guard;
+    {
+        EXPECT_EQ(Guard::count(), 2);
+        // NOLINTJUSTIFICATION ctor and dtor side effects are tested
+        // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+        Guard movedGuard(std::move(guard));
+        EXPECT_EQ(Guard::count(), 3);
+    }
+    EXPECT_EQ(Guard::count(), 2);
+
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, assignmentDoesNotChangeLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1c04ac75-d47a-44da-b8dc-6f567a53d3fc");
+    Foo::reset();
+    Guard guard1;
+    Guard guard2;
+
+    EXPECT_EQ(Guard::count(), 3);
+    guard1 = guard2;
+    EXPECT_EQ(Guard::count(), 3);
+    guard1 = std::move(guard2);
+    EXPECT_EQ(Guard::count(), 3);
+
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, destructionAtZeroCountWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8b5a22a9-87bc-434b-9d07-9f3c20a6944e");
+    Foo::reset();
+    {
+        Guard guard;
+        auto& instance = Guard::instance();
+
+        // count is expected to be 2,
+        // we ignore the guard of g_instance by setting it to 1,
+        // hence when guard is destroyed the instance will be destroyed as well
+        auto oldCount = Guard::setCount(1);
+        EXPECT_EQ(oldCount, 2);
+
+        EXPECT_EQ(Foo::ctorCalled, 0);
+        EXPECT_EQ(Foo::dtorCalled, 0);
+        EXPECT_EQ(instance.id, 1);
+    }
+
+    EXPECT_EQ(Guard::count(), 0);
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 1);
+}
+
+TEST_F(StaticLifetimeGuard_test, constructionAfterDestructionWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "0077e73d-ddf5-47e7-a7c6-93819f376175");
+    Foo::reset();
+    EXPECT_EQ(Guard::count(), 0);
+    {
+        Guard guard;
+        auto& instance = Guard::instance();
```

Comment:
As is, this test relies on destruction in the previous test. This is bad. It should take care of destruction before itself if needed. Will add code for that.

Reply:
Added code to ensure an existing instance is destroyed